taranis/clustering.py

```diff
@@ -0,0 +1,158 @@
+
+        Returns:
+            dict: where key is the cluster number and value is the list of
+                alleles belongs to the cluster
+        """
+        out_clusters = {}
+        for cluster_id in range(np.max(cluster_ids) + 1):
+            out_clusters[cluster_id] = [
+                id_to_seq_name[seq_id]
+                for seq_id in np.argwhere(cluster_ids == cluster_id).flatten()
+            ]
+
+        return out_clusters
+
+    def collect_data_cluster(self, src_cluster_ptrs: np.ndarray) -> dict:
+        """Collect the mean, index allele center and number of alleles in
+            cluster for each cluster
+
+        Args:
+            src_cluster_ptrs (np.ndarray): cluster matrix
+
+        Returns:
+            dict: where key is the cluster number and value a list of the
+                statistic data
+        """
+        log.debug(f"Collecting data for cluster {self.ref_seq_name}")
+        cluster_data = {}
+        for cluster_id in range(np.max(src_cluster_ptrs) + 1):
+            cluster_data[cluster_id] = {}
+            log.debug(f"calculating mean for cluster number {cluster_id}")
+            cluster_bool_ptrs = src_cluster_ptrs == cluster_id
+            cluster_mtrx_idxs = np.ix_(cluster_bool_ptrs, cluster_bool_ptrs)
+            row_idx_pos = np.argwhere(cluster_bool_ptrs).flatten()
+            # col_idx_pos = np.argwhere(cluster_bool_ptrs).flatten()
+            cluster_mean = self.calculate_mean_cluster(cluster_mtrx_idxs, row_idx_pos)
+            # get the closest distance coordenates to cluster mean value
+            cluster_data[cluster_id]["avg"] = cluster_mean
+            cluster_data[cluster_id]["center_id"] = self.calculate_cluster_center(
+                cluster_mtrx_idxs, cluster_mean
+            )
+            log.debug(f"Get the cluster center for {cluster_id}")
+            # get the number of sequences for the cluster
+            cluster_data[cluster_id]["n_seq"] = len(cluster_mtrx_idxs[0])
+        return cluster_data
+
+    def create_clusters(self) -> list[dict]:
+        """main method to create clustering using the Leiden algorithm
+
+        Returns:
+            list: two dictionaries are returned first with the cluster and the
+            matrix indexes adn second the statistics data for each cluster
+        """
+        comm_graph = ig.Graph.Weighted_Adjacency(
+            self.dist_matrix.tolist(), mode=1, loops=False
+        )
+        graph_clusters = leidenalg.find_partition(
+            comm_graph,
+            leidenalg.CPMVertexPartition,
+            weights="weight",
+            n_iterations=-1,
+            resolution_parameter=self.res_param,
+            seed=self.seed,
+        )
+        cluster_ptrs = np.array(graph_clusters.membership)
+
+        clusters_data = self.collect_data_cluster(cluster_ptrs)
+        # check that cluste average values are upper than 0.9
+        for value in clusters_data.values():
+            if value["avg"] < 0.9:
```

Comment:
What is this check for? Which is the cluster average value?

Reply:
This was for testing propose, to see if any of the cluster had and average bellow 0.9. However during design/testing I  never get an average under 0.9, So I will remove this piece of code, as it is not needed.